rustls/src/builder.rs

```diff
@@ -6,34 +6,101 @@ use crate::versions;
 use core::fmt;
 use core::marker::PhantomData;
 
-/// Building a [`ServerConfig`] or [`ClientConfig`] in a linker-friendly and
-/// complete way.
+/// A [builder] for [`ServerConfig`] or [`ClientConfig`] values.
 ///
-/// Linker-friendly: meaning unused cipher suites, protocol
-/// versions, key exchange mechanisms, etc. can be discarded
-/// by the linker as they'll be unreferenced.
+/// To get one of these, call [`ServerConfig::builder()`] or [`ClientConfig::builder()`].
 ///
-/// Complete: the type system ensures all decisions required to run a
-/// server or client have been made by the time the process finishes.
+/// To build a config, you must make at least three decisions (in order):
 ///
-/// Example, to make a [`ServerConfig`]:
+/// - Which protocol primitives should be supported (cipher suites, key exchange groups, protocol versions)?
+/// - How should this client or server verify certificates provided by its peer?
+/// - What certificates should this client or server present to its peer?
+///
+/// Once the ServerConfig or ClientConfig is built, there are additional configuration options
+/// you can set directly by setting fields on those objects. See [`ServerConfig`] and [`ClientConfig`] for details.
+///
+/// The usual choice for protocol primitives is to call
+/// [`ConfigBuilder::with_safe_defaults`], which will choose rustls' defaults for cipher suites, key
+/// exchange groups and protocol versions:
+///
+/// ```
+/// use rustls::{ClientConfig, ServerConfig, crypto::ring::Ring};
+/// // <Ring> specifies the cryptographic provider to use.
+/// ClientConfig::<Ring>::builder()
+///     .with_safe_defaults()
+/// //  ...
+/// # ;
+///
+/// ServerConfig::<Ring>::builder()
+///     .with_safe_defaults()
+/// //  ...
+/// # ;
+/// ```
```

Comment:
How about moving this section explaining that the usual choice is `with_safe_defaults` to be immediately after the description of the three choices to be made, moving the "additional configuration options" to after the safe defaults discussion?

If I'm someone reading these docs looking for the fastest route from docs -> a working solution I think having the safe defaults path presented as soon as possible is helpful. WDYT?

Reply:
My goal with the "additional configuration options" verbiage is this: up front, we're telling people they "must" make 3 configuration decisions. I wanted to avoid seeming like those were the only pieces of configuration available, or leave them sifting for information about what else they can configure.

I wound up revising that paragraph to be shorter and say more of what I meant:

> For settings besides these, see the fields of [ServerConfig](https://github.com/rustls/rustls/pull/server/struct.ServerConfig.html) and [ClientConfig](https://github.com/rustls/rustls/pull/client/struct.ClientConfig.html).

Let me know what you think.

Reply:
The updated version seems better, thanks!